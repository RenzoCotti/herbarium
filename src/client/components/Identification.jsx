import React, { Component } from "react";
import "./style/Identification.css";
import { capitaliseString } from "../../utility/utility";

//where, overall appearance, details, season

class Identification extends Component {
  state = {};

  constructor(props) {
    super(props);

    //utility
    this.renderOptional = this.renderOptional.bind(this);

    this.renderGeneral = this.renderGeneral.bind(this);
    this.renderStem = this.renderStem.bind(this);
    this.renderLeaf = this.renderLeaf.bind(this);
    this.renderFlowers = this.renderFlowers.bind(this);
    this.renderFruits = this.renderFruits.bind(this);
  }

  //if obj isnt null, render opt if is not null, otherwise render obj
  renderOptional(obj, str, opt) {
    if (obj) {
      return (
        <tr>
          <td className="sub-title">{str}:</td>
          <td>{opt ? opt : obj}</td>
        </tr>
      );
    }
    return;
  }

  renderGeneral(plant) {
    return (
      <React.Fragment>
        <tr>
          <th className="title">General</th>
        </tr>
        <tr>
          <td className="sub-title">Type:</td>
          <td>{plant.typeOfPlant}</td>
        </tr>
        <tr>
          <td className="sub-title">Evergreen:</td>
          <td> {plant.evergreen ? "Yes" : "No"}</td>
        </tr>
        <tr>
          <td className="sub-title">Zone:</td>
          <td>{plant.zone.join(", ")}</td>
        </tr>
        <tr>
          <td className="sub-title">Habitat:</td>
          <td>{plant.habitat}</td>
        </tr>
        <tr>
          <td className="sub-title">Height:</td>
          <td>{plant.height} m</td>
        </tr>
        {this.renderOptional(plant.description, "Description")}
      </React.Fragment>
    );
  }

  renderStem(plant) {
    return (
      <React.Fragment>
        <tr className="title">
          <th>Stem</th>
        </tr>
        <tr>
          <td className="sub-title">Colour:</td>
          <td>{plant.stemColour}</td>
        </tr>
        <tr>
          <td className="sub-title">Texture:</td>
          <td>{plant.stemTexture}</td>
        </tr>
        <tr>
          <td className="sub-title">Description:</td>
          <td>{plant.stemDescription}</td>
        </tr>
      </React.Fragment>
    );
  }

  renderLeaf(plant) {
    return (
      <React.Fragment>
        <tr className="title">
          <th>Leaves</th>
        </tr>
        <tr>
          <td className="sub-title">Shape:</td>
          <td>{plant.leafShape}</td>
        </tr>
        <tr>
          <td className="sub-title">Margin:</td>
          <td>{plant.leafMargin}</td>
        </tr>
        {this.renderOptional(plant.leafVenation, "Venation")}
        <tr>
          <td className="sub-title">Length:</td>
          <td>{plant.leafLength} cm</td>
        </tr>
        <tr>
          <td className="sub-title">Description:</td>
          <td>{plant.leafDescription}</td>
        </tr>
      </React.Fragment>
    );
  }

  renderFlowers(plant) {
    if (!plant.flowerColour) return;
    return (
      <React.Fragment>
        <tr className="title">
          <th>Flowers</th>
        </tr>
        {this.renderOptional(plant.flowerColour, "Colour")}
        {this.renderOptional(
          this.bloomMonth,
          "Blooming Months",
          plant.bloomMonth.join(", ")
        )}
        {this.renderOptional(plant.flowerDescription, "Description")}
      </React.Fragment>
    );
  }

  renderFruits(plant) {
    if (!plant.fruitColour) return;
    return (
      <React.Fragment>
        <tr className="title">
          <th>Fruit</th>
        </tr>
        {this.renderOptional(plant.fruitColour, "Colour")}
        {this.renderOptional(
          this.harvestMonth,
          "Harvest Months",
          plant.harvestMonth.join(", ")
        )}
        {this.renderOptional(plant.fruitDescription, "Description")}
        {this.renderOptional(plant.fruitSize, "Size", this.fruitSize + " cm")}
      </React.Fragment>
    );
  }

  render() {
    let plant = this.props.plant;
    return (
      <div className="identification">
        <div className="latin">{capitaliseString(plant.latinName)}</div>
        <div className="common">{capitaliseString(plant.commonName)}</div>
        <table>
          <tbody>
            {this.renderGeneral(plant)}
            {this.renderStem(plant)}
            {this.renderLeaf(plant)}
            {this.renderFlowers(plant)}
            {this.renderFruits(plant)}
          </tbody>
        </table>
      </div>
    );
  }
}

export default Identification;
