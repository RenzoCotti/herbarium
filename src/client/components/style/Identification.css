.title {
  margin-top: 15px;
  font-weight: bold;
}

.super-title {
  font-weight: bold;
  font-size: 24pt;
}

.identification {
  padding: 20px;
  background-color: #90ca40bb;
}

.latin {
  font-style: italic;
}

.common {
  font-size: 25pt;
  margin-bottom: 50px;
  font-weight: bold;
  color: darkgreen;
}
