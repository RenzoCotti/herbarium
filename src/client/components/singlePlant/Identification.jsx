import React, { Component } from "react";
import "../style/Identification.css";
import {
  capitalise,
  capitaliseString,
  renderOptional
} from "../../../utility/utility";

//where, overall appearance, details, season

class Identification extends Component {
  state = {};

  constructor(props) {
    super(props);
    this.renderGeneral = this.renderGeneral.bind(this);
    this.renderStem = this.renderStem.bind(this);
    this.renderLeaf = this.renderLeaf.bind(this);
    this.renderFlowers = this.renderFlowers.bind(this);
    this.renderFruits = this.renderFruits.bind(this);
  }

  renderGeneral(plant) {
    return (
      <React.Fragment>
        <tr>
          <th className="title">General</th>
        </tr>
        <tr>
          <td className="sub-title">Type:</td>
          <td>{capitalise(plant.typeOfPlant)}</td>
        </tr>
        <tr>
          <td className="sub-title">Evergreen:</td>
          <td>{plant.evergreen ? "Yes" : "No"}</td>
        </tr>
        <tr>
          <td className="sub-title">Regions:</td>
          <td>{plant.regions.join(", ")}</td>
        </tr>
        <tr>
          <td className="sub-title">Habitat:</td>
          <td>{capitalise(plant.habitat)}</td>
        </tr>
        <tr>
          <td className="sub-title">Height:</td>
          <td>{plant.height} m</td>
        </tr>
        {renderOptional(plant.description, "Description")}
      </React.Fragment>
    );
  }

  renderStem(plant) {
    return (
      <React.Fragment>
        <tr>
          <th className="title">Stem</th>
        </tr>
        <tr>
          <td className="sub-title">Colour:</td>
          <td>{capitalise(plant.stemColour)}</td>
        </tr>
        <tr>
          <td className="sub-title">Texture:</td>
          <td>{capitalise(plant.stemTexture)}</td>
        </tr>
        <tr>
          <td className="sub-title">Description:</td>
          <td>{capitalise(plant.stemDescription)}</td>
        </tr>
      </React.Fragment>
    );
  }

  renderLeaf(plant) {
    return (
      <React.Fragment>
        <tr>
          <th className="title">Leaves</th>
        </tr>
        <tr>
          <td className="sub-title">Shape:</td>
          <td>{capitalise(plant.leafShape)}</td>
        </tr>
        <tr>
          <td className="sub-title">Margin:</td>
          <td>{capitalise(plant.leafMargin)}</td>
        </tr>
        {renderOptional(plant.leafVenation, "Venation")}
        <tr>
          <td className="sub-title">Length:</td>
          <td>{plant.leafLength} cm</td>
        </tr>
        <tr>
          <td className="sub-title">Description:</td>
          <td>{capitalise(plant.leafDescription)}</td>
        </tr>
      </React.Fragment>
    );
  }

  renderFlowers(plant) {
    if (!plant.flowerColour) return;
    return (
      <React.Fragment>
        <tr>
          <th className="title">Flowers</th>
        </tr>
        {renderOptional(plant.flowerColour, "Colour")}
        {renderOptional(
          this.bloomMonth,
          "Blooming Months",
          plant.bloomMonth.join(", ")
        )}
        {renderOptional(plant.flowerDescription, "Description")}
      </React.Fragment>
    );
  }

  renderFruits(plant) {
    if (!plant.fruitColour) return;
    return (
      <React.Fragment>
        <tr>
          <th className="title">Fruit</th>
        </tr>
        {renderOptional(plant.fruitColour, "Colour")}
        {renderOptional(
          this.harvestMonth,
          "Harvest Months",
          plant.harvestMonth.join(", ")
        )}
        {renderOptional(plant.fruitDescription, "Description")}
        {renderOptional(plant.fruitSize, "Size", this.fruitSize + " cm")}
      </React.Fragment>
    );
  }

  render() {
    let plant = this.props.plant;

    return (
      <div className="secondary-container">
        <div className="latinName">{capitaliseString(plant.latinName)}</div>
        <div className="commonName">{capitaliseString(plant.commonName)}</div>
        <table>
          <tbody>
            {this.renderGeneral(plant)}
            {this.renderStem(plant)}
            {this.renderLeaf(plant)}
            {this.renderFlowers(plant)}
            {this.renderFruits(plant)}
          </tbody>
        </table>
      </div>
    );
  }
}

export default Identification;
