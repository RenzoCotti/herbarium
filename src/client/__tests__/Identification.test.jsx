import Identification from "../components/Identification";
import React from "react";
import { shallow } from "../enzyme";
let mint = require("../../../content")[1];

describe("Identification", () => {
  it("renders correctly", () => {
    const identify = shallow(<Identification plant={mint} />);
    expect(identify).toMatchSnapshot();
  });
  it("should have 3 children", () => {
    const identify = shallow(<Identification plant={mint} />);
    expect(identify.children()).toHaveLength(3);
  });
  it("latin should be equal", () => {
    const identify = shallow(<Identification plant={mint} />);
    expect(identify.find(".latin").text()).toBe(mint.latin);
  });

  it("name should be equal", () => {
    const identify = shallow(<Identification plant={mint} />);
    expect(identify.find(".common").text()).toBe(mint.common);
  });
});
