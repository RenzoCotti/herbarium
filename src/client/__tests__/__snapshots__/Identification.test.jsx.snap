// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Identification renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Identification
    plant={
      Object {
        "blooming": Array [
          "mid summer",
          "late summer",
        ],
        "common": "Peppermint",
        "flowers": Array [
          "purple",
        ],
        "habitat": "gravel",
        "harvest": "spring",
        "height": 0.6,
        "images": Object {
          "flowers": "https://cdn.shopify.com/s/files/1/1766/2959/products/Peppermint_4_400x400.jpg?v=1535245781",
          "leaves": "https://images-na.ssl-images-amazon.com/images/I/51TeuM4HqJL.jpg",
          "overall": "https://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Mentha-piperita.JPG/330px-Mentha-piperita.JPG",
        },
        "latin": "Mentha Piperita",
        "leaves": Object {
          "height": 2.5,
          "margin": "toothed",
          "properties": Array [
            "dark green",
            "red",
          ],
          "shape": "lanceolate",
          "width": 6,
        },
        "region": "Tundra",
        "shape": "bush",
        "stem": "smooth",
        "type": "herbaceous plant",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="latin"
        >
          Mentha Piperita
        </div>,
        <div
          className="common"
        >
          Peppermint
        </div>,
        <div>
          <div
            className="title"
          >
            Stem
          </div>
          <div>
            smooth
          </div>
          <div
            className="title"
          >
            Leaves
          </div>
          <div>
            Shape: 
            lanceolate
          </div>
          <div>
            Properties: 
            dark green red
          </div>
          <div>
            Margin: 
            toothed
          </div>
          <div>
            Size(W x H): 
            6
            cm x
            2.5
            cm
          </div>
          <div
            className="title"
          >
            Flowers
          </div>
          <div>
            purple
          </div>
        </div>,
      ],
      "className": "identification",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Mentha Piperita",
          "className": "latin",
        },
        "ref": null,
        "rendered": "Mentha Piperita",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Peppermint",
          "className": "common",
        },
        "ref": null,
        "rendered": "Peppermint",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="title"
            >
              Stem
            </div>,
            <div>
              smooth
            </div>,
            <div
              className="title"
            >
              Leaves
            </div>,
            <div>
              Shape: 
              lanceolate
            </div>,
            <div>
              Properties: 
              dark green red
            </div>,
            <div>
              Margin: 
              toothed
            </div>,
            <div>
              Size(W x H): 
              6
              cm x
              2.5
              cm
            </div>,
            <div
              className="title"
            >
              Flowers
            </div>,
            <div>
              purple
            </div>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Stem",
              "className": "title",
            },
            "ref": null,
            "rendered": "Stem",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "smooth",
            },
            "ref": null,
            "rendered": "smooth",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Leaves",
              "className": "title",
            },
            "ref": null,
            "rendered": "Leaves",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Shape: ",
                "lanceolate",
              ],
            },
            "ref": null,
            "rendered": Array [
              "Shape: ",
              "lanceolate",
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Properties: ",
                "dark green red",
              ],
            },
            "ref": null,
            "rendered": Array [
              "Properties: ",
              "dark green red",
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Margin: ",
                "toothed",
              ],
            },
            "ref": null,
            "rendered": Array [
              "Margin: ",
              "toothed",
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Size(W x H): ",
                6,
                "cm x",
                2.5,
                "cm",
              ],
            },
            "ref": null,
            "rendered": Array [
              "Size(W x H): ",
              6,
              "cm x",
              2.5,
              "cm",
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Flowers",
              "className": "title",
            },
            "ref": null,
            "rendered": "Flowers",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "purple",
            },
            "ref": null,
            "rendered": "purple",
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="latin"
          >
            Mentha Piperita
          </div>,
          <div
            className="common"
          >
            Peppermint
          </div>,
          <div>
            <div
              className="title"
            >
              Stem
            </div>
            <div>
              smooth
            </div>
            <div
              className="title"
            >
              Leaves
            </div>
            <div>
              Shape: 
              lanceolate
            </div>
            <div>
              Properties: 
              dark green red
            </div>
            <div>
              Margin: 
              toothed
            </div>
            <div>
              Size(W x H): 
              6
              cm x
              2.5
              cm
            </div>
            <div
              className="title"
            >
              Flowers
            </div>
            <div>
              purple
            </div>
          </div>,
        ],
        "className": "identification",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Mentha Piperita",
            "className": "latin",
          },
          "ref": null,
          "rendered": "Mentha Piperita",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Peppermint",
            "className": "common",
          },
          "ref": null,
          "rendered": "Peppermint",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="title"
              >
                Stem
              </div>,
              <div>
                smooth
              </div>,
              <div
                className="title"
              >
                Leaves
              </div>,
              <div>
                Shape: 
                lanceolate
              </div>,
              <div>
                Properties: 
                dark green red
              </div>,
              <div>
                Margin: 
                toothed
              </div>,
              <div>
                Size(W x H): 
                6
                cm x
                2.5
                cm
              </div>,
              <div
                className="title"
              >
                Flowers
              </div>,
              <div>
                purple
              </div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Stem",
                "className": "title",
              },
              "ref": null,
              "rendered": "Stem",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "smooth",
              },
              "ref": null,
              "rendered": "smooth",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Leaves",
                "className": "title",
              },
              "ref": null,
              "rendered": "Leaves",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Shape: ",
                  "lanceolate",
                ],
              },
              "ref": null,
              "rendered": Array [
                "Shape: ",
                "lanceolate",
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Properties: ",
                  "dark green red",
                ],
              },
              "ref": null,
              "rendered": Array [
                "Properties: ",
                "dark green red",
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Margin: ",
                  "toothed",
                ],
              },
              "ref": null,
              "rendered": Array [
                "Margin: ",
                "toothed",
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Size(W x H): ",
                  6,
                  "cm x",
                  2.5,
                  "cm",
                ],
              },
              "ref": null,
              "rendered": Array [
                "Size(W x H): ",
                6,
                "cm x",
                2.5,
                "cm",
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Flowers",
                "className": "title",
              },
              "ref": null,
              "rendered": "Flowers",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "purple",
              },
              "ref": null,
              "rendered": "purple",
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
